core/src/family.cc

```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
 
 template <typename T>
 std::vector<MetricFamily> Family<T>::Collect() {
-  std::lock_guard<std::mutex> lock{mutex_};
   auto family = MetricFamily{};
-  family.name = name_;
-  family.help = help_;
-  family.type = T::metric_type;
-  for (const auto& m : metrics_) {
-    family.metric.push_back(std::move(CollectMetric(m.first, m.second.get())));
+  std::vector<std::shared_ptr<T>> to_be_removed;
+  {
+    std::lock_guard<std::mutex> lock{mutex_};
+    family.name = name_;
+    family.help = help_;
+    family.type = T::metric_type;
+    for (const auto& m : metrics_) {
+      if (!m.second->Expired()) {
+        family.metric.push_back(std::move(CollectMetric(m.first, m.second)));
+      } else if (retention_behavior_ == RetentionBehavior::Remove) {
+        to_be_removed.push_back(m.second);
+      }
+    }
+  } // end mutex
+  for (auto metric: to_be_removed) {
+    Remove(metric);
+  }
+  if (family.metric.empty()) {
+    return {};
+  } else {
+    return {family};
+  }
+}
+
+template <typename T>
+bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name, const std::map<std::string, std::string>& re_labels, const bool& bump, const bool& debug) {
+  /* Setup */
+  bool modified(false);
+  const boost::regex name_expr(re_name);
+  if (debug) {  
+    printf("UpdateRetentionTime: %s \n", re_name.c_str());
+  }
+  /* Check if the name is a match */
+  if (boost::regex_match(name_, name_expr)) {
+    for (auto metric: metrics_) {
+      bool matched(true);
```

Comment:
Based on this logic, I'm not too sure why we have both `matched` and `found`.  I think it looks like you could just use `matched` and then set it to true  under the constant or metric label conditions.

Reply:
There is a bit of a bug in the logic, see followup comment.